### DISCUSSION TITLE
Question on filtering grains and difficulty computing twin length fraction and taylor factor

### DISCUSSION THREAD

**Discussion by StopkaKris** [mtex-toolbox/mtex#1452] [Ask Anything] [answered]
Hello,

I have an EBSD data set for Ni-base superalloy IN718 (fcc crystal structure) that contains Σ3 twin (i.e., nominal grain misorientation of 60° about the <111> axis) and I would like to characterize the following:

- Number of grains and twins
- Mean grain size and standard deviation in microns, including twins in this calculation
- Both the percent of grain boundaries that are twin boundaries, AND the twin length fraction, i.e., the ratio of the sum of twin boundary lengths to the total sum of grain boundary lengths.
- Taylor factor of each grain and twin

I would like to compute these values after specifying a minimum grain size of 12 pixels. I put together the code below that gives me the first two bullets above:

```
%% Specify Crystal and Specimen Symmetries
% crystal symmetry

clear all
close all
clc

CS = {... 
  'notIndexed',...
  crystalSymmetry('m-3m', [3.6 3.6 3.6], 'mineral', 'Ni-superalloy', 'color', [0.53 0.81 0.98])};

% plotting convention
setMTEXpref('xAxisDirection','north');
setMTEXpref('zAxisDirection','outOfPlane');

%% Specify File Names

% path to files
pname = 'C:\...\NIST-DOE-8-65_V2';

% which files to be imported
fname = [pname '\NIST-DOE-8-65_V2.ctf'];

%% Import the Data

% create an EBSD variable containing the data
ebsd = EBSD.load(fname,CS,'interface','ctf',...
  'convertEuler2SpatialReferenceFrame');

% Compute grains with misorientation tolerance of 5 degrees
[grains,ebsd.grainId] = calcGrains(ebsd('indexed'),'angle',5*degree);

% Remove grains with less than some number of pixels
ebsd(grains(grains.grainSize < 11)) = [];

% and repeat the grain computation
[grains,ebsd.grainId] = calcGrains(ebsd('indexed'),'angle',5*degree);

gB = grains.boundary(); 

% Compute percent of twin boundaries with tolerance of 1 degree
CSL_3_GB_percentage = 100 * volume(gB.misorientation,CSL(3,ebsd.CS),1*degree);

% Compute equivalent grain diameter of each grain
grain_sizes = sqrt(area(grains)/pi)*2;

fprintf('\n')
fprintf('Number of grains: %d\n', length(grain_sizes))
fprintf('Mean and std dev of grains: %0.3f, %0.3f\n', mean(grain_sizes), std(grain_sizes))
fprintf('Percent of boundaries as twin boundaries: %0.3f%%\n', CSL_3_GB_percentage)
fprintf('\n')
```

- Am I correctly filtering out grains less than 12 pixels in size with the line `ebsd(grains(grains.grainSize < 11)) = [];`? Does MTEX then "grow" the surrounding grains to fill the space, or does it just ignore the pixels that these grains comprised? If it is the latter, is there a way to make MTEX isotropically grow the surrounding regions after filtering out grains smaller than 12 pixels?
- What command do I use to sum all grain boundaries and then only the twin boundaries? I believe `sum(grains.boundary('indexed').segLength)` will give me the sum of all grain boundary lengths, correct?
- Am I correct in understanding that the variable `CSL_3_GB_percentage` as computed above is the PERCENT of twin boundaries, but not the LENGTH FRACTION of twin boundaries?

Finally, I appended the code below to compute the taylor factor for each grain and twin

```% some strain
q = 0;
epsilon = strainTensor(diag([1 -q -(1-q)]))

% consider fcc slip systems
sS = symmetrise(slipSystem.fcc(grains.CS));

% apply Taylor model
[M,b,W] = calcTaylor(inv(grains.meanOrientation)*epsilon,sS);
```

but the call to the function `calcTaylor` gives the following error:
``` 
Error using optimoptions (line 124)
Empty keys are not allowed in this container.

Error in strainTensor/calcTaylor (line 65)
  options = optimoptions('linprog','Algorithm','interior-point-legacy','Display','none');
```
Thank you in advance for your help!

* EDIT 
I realize that the last issue with regard to computing the Taylor factor is because I did not have the global optimization toolbox installed. The  last bit of code works after the toolbox is installed!

* EDIT 2
When I use `plot(ebsd, ebsd.orientations)` to visualize the data, I see many red pixels that I assume are not included in the calculations above. I assume this is because the image quality of these pixels is low and MTEX considers these as "bad data" and therefore does not consider them in the calculation, correct? Is there a way to both remove these and then dilate/isotropically grow the neighboring grains and twins, so that my metrics of interest above utilize the entire EBSD scan area?
I attempted to "fill" in these pixels using the updated code below but when I plot the ebsd map, I still see the red pixels.

```
% compute grains
[grains,ebsd.grainId] = calcGrains(ebsd('indexed'),'angle',5*degree);

% remove small grains
ebsd(grains(grains.grainSize < 12)) = [];

% compute grains
[grains,ebsd.grainId] = calcGrains(ebsd('indexed'),'angle',5*degree);

%% Fill in grains !!
ebsd = fill(ebsd);

% and repeat the grain computation
[grains,ebsd.grainId] = calcGrains(ebsd('indexed'),'angle',5*degree);
```

Best regards,
Kris

**Comment by kilir** [comment, marked as answer]
Hi,

> Am I correctly filtering out grains less than 12 pixels in size with the line ebsd(grains(grains.grainSize < 11)) = [];

No, it removes the EBSD point which belong to grains with less than 11 points

> Does MTEX then "grow" the surrounding grains to fill the space, or does it just ignore the pixels that these grains comprised?

Depends on how you compute grains, see here: https://mtex-toolbox.github.io/GrainReconstructionAdvanced.html

>  If it is the latter, is there a way to make MTEX isotropically grow the surrounding regions after filtering out grains smaller than 12 pixels?

And stop once the shortest distance to a populated neighboring region is reached? No

>  I believe sum(grains.boundary('indexed').segLength) will give me the sum of all grain boundary lengths, correct?

yes

> Am I correct in understanding that the variable CSL_3_GB_percentage as computed above is the PERCENT of twin boundaries, but not the LENGTH FRACTION of twin boundaries?
`CSL_3_GB_percentage = 100 * volume(gB.misorientation,CSL(3,ebsd.CS),1*degree);`

 No, this is the volume in misorienation space occupied by misorientations which deviate at the most be 1 degree from a CSL3

> When I use plot(ebsd, ebsd.orientations) to visualize the data, I see many red pixels that I assume are not included in the calculations above. I assume this is because the image quality of these pixels is low and MTEX considers these as "bad data" and therefore does not consider them in the calculation, correct? 

No, this is not correct, it just means those grains have orientations with the c-axis being parallel to z in specimen coordinates, there is no automagic discarding of "bad" data. If you think you want to remove points which belong to a certain orientation, you , e.g. have a look here: https://mtex-toolbox.github.io/EBSD.findByOrientation.html if you just want to remove grains which only have one direction parallel to a ref. direction, use `angle()`.

Hope this helps.

Cheers,
Rüdiger

Maybe it's easier to digest if you create for each of your future questions a separate thread.

**Reply by StopkaKris** [reply]
Hi Rüdiger,

Thank you for your response and I will definitely separate my questions into different threads in the future. And yes, this helps a ton! I previously tried to perform all these calculations in DREAM.3D but wanted to cross validate my results. I did have some follow up questions, however:

- Ah, I see that I mistakenly wrote 11 instead of 12 in the line `ebsd(grains(grains.grainSize < 11)) = [];`
- To be more specific, the variable `CSL_3_GB_percentage` is the _volume fraction_ of misorientations which deviate a maximum of 1 degree from a CSL3 boundary, correct? I just want to make sure I fully understand this...
- Another calculation I think I now understand is how to compute the fraction of grain boundaries that are within 1 degree of a CSL3 boundary. Can you please verify this? I simply sum the gb lengths for CSL3 and divide by the total gb length in the EBSD scan.
```
% Sum of all GBs
sum_all_GBs = sum(grains.boundary('indexed').segLength);

% select CSL(3) grain boundaries
gB3 = gB(angle(gB.misorientation,CSL(3,ebsd.CS)) < 1*degree);
sum_twin_GBs = sum(gB3.segLength);

twin_length_fraction = sum_twin_GBs / sum_all_GBs;
```

- I am also interested in statistics regarding the merged (i.e., parent) grains after twins are "absorbed". I believe this is simply the code below, correct?
```
% Get info of parent grains
[mergedGrains,parentId] = merge(grains,gB3);
grain_sizes_parent_grains = sqrt(area(mergedGrains)/pi)*2;
```

Best regards,
Kris

**Reply by kilir** [reply]
Hi

> To be more specific, the variable CSL_3_GB_percentage is the volume fraction of misorientations which deviate a maximum of 1 degree from a CSL3 boundary, correct? I just want to make sure I fully understand this...

Actually in you case for discrete points it returns what you are looking for, the fraction of boundries which fulfill the condition relative to the the total number. I was thinking on odfs. You are right, here it is equivalent to the (unweighted) number fraction.

>  Can you please verify this? I simply sum the gb lengths for CSL3 and divide by the total gb length in the EBSD scan.

This is correct

> I am also interested in statistics regarding the merged (i.e., parent) grains after twins are "absorbed". I believe this is simply the code below, correct?

yes, this is the equivalent diameter of all grains after merging (including those which did not contain a CSL3)
you can also use `2* mergedGrains.equivalentRadius`.

Cheers,
Rüdiger

**Reply by StopkaKris** [reply]
Great, thanks so much, Rüdiger!
